File: src/bun.js/test/expect/toHaveLastReturnedWith.zig
pub fn toHaveLastReturnedWith(this: *Expect, globalThis: *JSGlobalObject, callframe: *CallFrame) bun.JSError!JSValue {
    jsc.markBinding(@src());

    const thisValue = callframe.this();
    defer this.postMatch(globalThis);

    const value: JSValue = try this.getValue(globalThis, thisValue, "toHaveBeenLastReturnedWith", "<green>expected<r>");

    const expected = callframe.argumentsAsArray(1)[0];
    this.incrementExpectCallCounter();

    const returns = try bun.cpp.JSMockFunction__getReturns(globalThis, value);
    if (!returns.jsType().isArray()) {
        var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        defer formatter.deinit();
        return globalThis.throw("Expected value must be a mock function: {f}", .{value.toFmt(&formatter)});
    }

    const calls_count = @as(u32, @intCast(try returns.getLength(globalThis)));
    var pass = false;
    var last_return_value: JSValue = .js_undefined;
    var last_call_threw = false;
    var last_error_value: JSValue = .js_undefined;

    if (calls_count > 0) {
        const last_result = returns.getDirectIndex(globalThis, calls_count - 1);

        if (last_result.isObject()) {
            const result_type = try last_result.get(globalThis, "type") orelse .js_undefined;
            if (result_type.isString()) {
                const type_str = try result_type.toBunString(globalThis);
                defer type_str.deref();

                if (type_str.eqlComptime("return")) {
                    last_return_value = try last_result.get(globalThis, "value") orelse .js_undefined;

                    if (try last_return_value.jestDeepEquals(expected, globalThis)) {
                        pass = true;
                    }
                } else if (type_str.eqlComptime("throw")) {
                    last_call_threw = true;
                    last_error_value = try last_result.get(globalThis, "value") orelse .js_undefined;
                }
            }
        }
    }

    if (pass != this.flags.not) {
        return .js_undefined;
    }

    // Handle failure
    var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
    defer formatter.deinit();

    const signature = comptime getSignature("toHaveBeenLastReturnedWith", "<green>expected<r>", false);

    if (this.flags.not) {
        return this.throw(globalThis, comptime getSignature("toHaveBeenLastReturnedWith", "<green>expected<r>", true), "\n\n" ++ "Expected mock function not to have last returned: <green>{f}<r>\n" ++ "But it did.\n", .{expected.toFmt(&formatter)});
    }

    if (calls_count == 0) {
        return this.throw(globalThis, signature, "\n\n" ++ "The mock function was not called.", .{});
    }

    if (last_call_threw) {
        return this.throw(globalThis, signature, "\n\n" ++ "The last call threw an error: <red>{f}<r>\n", .{last_error_value.toFmt(&formatter)});
    }

    // Diff if possible
    if (expected.isString() and last_return_value.isString()) {
        const diff_format = DiffFormatter{ .expected = expected, .received = last_return_value, .globalThis = globalThis, .not = false };
        return this.throw(globalThis, signature, "\n\n{f}\n", .{diff_format});
    }

    return this.throw(globalThis, signature, "\n\nExpected: <green>{f}<r>\nReceived: <red>{f}<r>", .{ expected.toFmt(&formatter), last_return_value.toFmt(&formatter) });
}

const bun = @import("bun");
const DiffFormatter = @import("../diff_format.zig").DiffFormatter;

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;
const mock = bun.jsc.Expect.mock;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/runtime/test_runner/expect/toHaveLastReturnedWith.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};

use super::mock;
use super::Expect;
use super::throw;

pub(crate) fn to_have_last_returned_with(
    this: &Expect,
    global_this: &JSGlobalObject,
    callframe: &CallFrame,
) -> JsResult<JSValue> {
    bun_jsc::mark_binding!();
    let expected = callframe.arguments_as_array::<1>()[0];
    let (this, returns, _value) = this.mock_prologue(
        global_this,
        callframe.this(),
        "toHaveBeenLastReturnedWith",
        "<green>expected<r>",
        mock::MockKind::Returns,
    )?;

    let calls_count = u32::try_from(returns.get_length(global_this)?).unwrap();
    let mock::ReturnCheck {
        matched: pass,
        return_value: last_return_value,
        threw: last_call_threw,
        error_value: last_error_value,
    } = if calls_count > 0 {
        mock::check_returned_with(
            global_this,
            returns.get_direct_index(global_this, calls_count - 1)?,
            expected,
        )?
    } else {
        mock::ReturnCheck::MISSING
    };

    if pass != this.flags.get().not() {
        return Ok(JSValue::UNDEFINED);
    }

    // Handle failure
    let signature = Expect::get_signature("toHaveBeenLastReturnedWith", "<green>expected<r>", false);

    if this.flags.get().not() {
        return mock::throw_not_failure(
            &this,
            global_this,
            "toHaveBeenLastReturnedWith",
            "<green>expected<r>",
            format_args!("Expected mock function not to have last returned"),
            expected,
            "\nBut it did.\n",
        );
    }

    if calls_count == 0 {
        return throw!(this, global_this, signature, "\n\nThe mock function was not called.");
    }

    if last_call_threw {
        return mock::throw_call_threw(
            &this,
            global_this,
            signature,
            format_args!("The last call"),
            last_error_value,
        );
    }

    // Diff if possible
    if expected.is_string() && last_return_value.is_string() {
        return mock::throw_diff(
            &this,
            global_this,
            signature,
            format_args!(""),
            expected,
            last_return_value,
        );
    }

    mock::throw_expected_received(
        &this,
        global_this,
        signature,
        format_args!(""),
        expected,
        last_return_value,
    )
}
